Fixed #26322 -- Updated lazy model reference section of fields docs. (#18675)
django/django: docs/ref/models/fields.txt

A many-to-one relationship. Requires two positional arguments: the class to
which the model is related and the :attr:`~ForeignKey.on_delete` option.

The first positional argument can be either a concrete model class or a
:ref:`lazy reference <lazy-relationships>` to a model class.
:ref:`Recursive relationships <recursive-relationships>`, where a model has a
relationship with itself, are also supported.

See :attr:`ForeignKey.on_delete` for details on the second positional
argument.

Database Representation
~~~~~~~~~~~~~~~~~~~~~~~

Behind the scenes, Django appends ``"_id"`` to the field name to create its

> database column name. In the above example, the database table for the ``Car``

Reviewer: With the extracted code, this phrase is now confusing, because it references a `Car` class that hasn't been created nor discussed so far. Would you have ideas in how to approach this?

One initial option I could think of is leaving an example in place for using an FK with a `Car` model, but without using any sort of recursion nor lazy reference.

Author: Good catch! I missed that

model will have a ``manufacturer_id`` column. (You can change this explicitly by
specifying :attr:`~Field.db_column`.) However, your code should never have to
deal with the database column name, unless you write custom SQL. You'll always
deal with the field names of your model object.

> A database index is automatically created on the ``ForeignKey``. You can

Author: Moved this here for better flow in ForeignKey's intro

A many-to-one relationship. Requires two positional arguments: the class to
which the model is related and the :attr:`~ForeignKey.on_delete` option.

The first positional argument can be either a concrete model class or a
:ref:`lazy reference <lazy-relationships>` to a model class.
:ref:`Recursive relationships <recursive-relationships>`, where a model has a
relationship with itself, are also supported.

See :attr:`ForeignKey.on_delete` for details on the second positional
argument.

Database Representation
~~~~~~~~~~~~~~~~~~~~~~~

Behind the scenes, Django appends ``"_id"`` to the field name to create its
database column name. For example, if a foreign key field is defined like::

>     manufacturer = models.ForeignKey(Manufacturer, on_delete=models.CASCADE)

Reviewer: Addendum: I worked on this while making my mind on the best solution, so I will push some minor edits.

the database table will have a ``manufacturer_id`` column. (You can change
this explicitly by specifying :attr:`~Field.db_column`.) However, your code
should never have to deal with the database column name, unless you write
custom SQL. You'll always deal with the field names of your model object.

> A database index is automatically created on the ``ForeignKey``. You can

Reviewer: Addendum: I see from your commit message that this change was intentional. I feel is a bit out of scope from this ticket, so I reverted this bit along with other small changes that I'm pushing.

> A database index is automatically created on the ``ForeignKey``. You can

Author: Looks good! Thanks

See :doc:`One-to-one relationships </topics/db/examples/one_to_one>` for usage
examples of ``OneToOneField``.

.. _lazy-relationships:

Lazy relationships
------------------

> Lazy relationships allow you to reference models using their names as strings
> or to create recursive relationships. You can use strings as the first
> argument in any relationship field to reference models lazily. The following
> types of lazy references are available:

Reviewer: In general, I tend to suggest to avoid this type of constructs (using "you"). I think the following is a bit more clear or professional:
```suggestion
Lazy relationships allow referencing models by their names (as strings) or
creating recursive relationships. Strings can be used as the first argument in
any relationship field to reference models lazily.

A lazy references can be either recursive, relative or absolute.
```
(We should definitely add links to each reference type.)

Author: So do I, but because Django docs use the second POV so much (sometimes informally), I am always unsure whether to use the professional tone (third POV), or else to preserve consistency and write in the same tone used in surrounding docs. Maybe we ought to include something in the style guides.

I applied the changes here though.

.. _recursive-relationships:

Recursive
~~~~~~~~~

To define a relationship where a model references itself, use ``"self"`` as
the first argument of the relationship field::

    from django.db import models

    class Manufacturer(models.Model):
        suppliers = models.ManyToManyField("self", symmetrical=False)

When used in an :ref:`abstract model <abstract-base-classes>`, the recursive
relationship will point to concrete model subclasses.

Relative
~~~~~~~~

If you need to create a relationship on a model that has not yet been defined,
you can use the name of the model instead of the model object itself::

    from django.db import models

    class Car(models.Model):
        manufacturer = models.ForeignKey(
            "Manufacturer",
            on_delete=models.CASCADE,
        )
        # ...

    class Manufacturer(models.Model):
        # ...

>         pass

Reviewer: I know this comes from the old content, but I would advice to use the same class definition as the one in `Recursive`, to avoid confusing the reader.

Relationships defined this way on :ref:`abstract models
<abstract-base-classes>` are resolved when the model is subclassed as a
concrete model and are not relative to the abstract model's ``app_label``:

.. code-block:: python
    :caption: ``products/models.py``

    from django.db import models

    class AbstractCar(models.Model):
        manufacturer = models.ForeignKey("Manufacturer", on_delete=models.CASCADE)

        class Meta:
            abstract = True

.. code-block:: python
    :caption: ``production/models.py``

    from django.db import models
    from products.models import AbstractCar

    class Manufacturer(models.Model):
        pass

    class Car(AbstractCar):
        pass

    # Car.manufacturer will point to `production.Manufacturer` here.

> .. admonition:: Blueprint models with relative references

Author: Added this as per @charettes's [comment](https://github.com/django/django/pull/6178#discussion_r54976680).

Reviewer: Thank you for working on this @cliff688! For me (non native speaker), reading "blueprint models" raises an eyebrow because I'm not really sure what it means. Perhaps:
```suggestion
.. admonition:: Reusable models with relative references
```

Author: Yeah. I think if we can remove or explain "blueprint" in the note itself, then I believe we will have made the meaning clearer. Currently, removing or replacing it with "reusable" does not say anything especially noteworthy of relative references.

Reviewer: I see your point @cliff688, thanks for raising that.

One idea would be to provide a small example within this subsection to illustrate this admonition. This would mean that the admonition title gets bumped to a subsubsection (is it possible?) and we provide a concise example. What I don't like about this approach is that this part of the docs is about "references", i.e. study material when following the [Diataxis model](https://diataxis.fr/). According to this, examples and "recipes" should be in the How-To section instead, because those are practical.

So, in summary, I'm not 100% sure of the best course of action. I think that an acceptable middle ground may be the upgrading of the admonition to a entry on its own with the smallest (perhaps slightly incomplete) example along the commentary to explain what a blueprint model is? What do you think?

(In the mid term, one could evaluate a separated, detailed How-To and point to that from here, but I would say that is outside the scope of this ticket.)

Author: Thank you @nessita for link to Diataxis! I enjoyed reading it.

About the admonition, I'm not in favour of adding an example here or in the "How-to" docs. Here's the admonition as it currently stands:
```
.. admonition:: Reusable models with relative references

    Relative references allow the creation of reusable abstract models
    with relationships that can resolve to different implementations
    of the referenced models in various subclasses across different
    applications.
```
Firstly, adding an example might overcomplicate the explanation, especially as the use case is more niche and abstract than common.

Secondly, I think as it currently stands, the admonition is explained in a way that should be clear to the target audience, and for those familiar with the concepts mentioned in the admonition, further illustration should be unnecessary.

Reviewer: @cliff688 Then maybe I misunderstood your previous comment. When you said:

> removing or replacing it with "reusable" does not say anything especially noteworthy of relative references.

I thought you were implying that we should add "something" to the current proposal to improve it. Could you expand on what did you mean please?

Author: Oh, I see 👍🏾 When I made that comment, the admonition read:
```
    Relative references enable the definition of reusable blueprint models with
    abstract classes, allowing relationships to resolve in concrete subclasses
    across different applications.
```
which after removing "blueprint" would not say anything that hadn't been said before (i.e. we already mentioned how relative references resolve in concrete model subclasses, and we already know that abstract models are reusable). However, I changed the admonition some moments afterwards (outdating the comment) to:
```
.. admonition:: Reusable models with relative references

    Relative references allow the creation of reusable abstract models
    with relationships that can resolve to different implementations
    of the referenced models in various subclasses across different
    applications.
```
which I think says something noteworthy about relative references in particular; namely, the bit about pointing to different implementations using the same relative reference. Also, with the new wording, we had dropped "blueprint".

Reviewer: Amazing, thank you @cliff688. I believe we are in agreement then, we'll keep the admonition as is.

I'll give a final review soon and merge afterwards, great work!

Author: Thank you @nessita 🌻

>     Relative references allow you to define reusable blueprint models
>     with abstract classes, as relationships can resolve in concrete
>     subclasses across different applications.

Reviewer: Thank you! May I suggest:
```suggestion
    Relative references enable the definition of reusable blueprint models with
    abstract classes, allowing relationships to resolve in concrete subclasses
    across different applications.
```

Author: I included these changes. Certainly an improvement on what was there before.

But I am not 100% satisfied that we're being as clear as we could be here. My understanding was that the suggestion meant for the docs to point out that because resolutions of relative references in different subclasses can point to different implementations of the pointed-to models, we can use them to create reusable "blueprint" models with relationships.

What do you think?

Absolute
~~~~~~~~

> Absolute references allow you to specify a model using its ``app_label``
> and class name, providing a way to reference models across different
> applications. This sort of lazy relationship can be useful in resolving
> circular imports.

Reviewer: How about:
```suggestion
Absolute references specify a model using its ``app_label`` and class name,
allowing for model references across different applications. This type of lazy
relationship can help resolve circular imports.
```

See :doc:`One-to-one relationships </topics/db/examples/one_to_one>` for usage
examples of ``OneToOneField``.

.. _lazy-relationships:

Lazy relationships
------------------

Lazy relationships allow referencing models by their names (as strings) or
creating recursive relationships. Strings can be used as the first argument in
any relationship field to reference models lazily. A lazy reference can be
either :ref:`recursive <recursive-relationships>`,
:ref:`relative <relative-relationships>` or
:ref:`absolute <absolute-relationships>`.

.. _recursive-relationships:

Recursive
~~~~~~~~~

To define a relationship where a model references itself, use ``"self"`` as the
first argument of the relationship field::

    from django.db import models

    class Manufacturer(models.Model):
        suppliers = models.ManyToManyField("self", symmetrical=False)
        # ...

When used in an :ref:abstract model <abstract-base-classes>, the recursive

> relationship resolves such that each concrete subclass references itself.

Author: @nessita, I reworded this a bit. While more verbose than the previous wording, I believe it's much clearer. Everything else looks great to me. Thank you 🌻

.. _relative-relationships:

Relative
~~~~~~~~

When a relationship needs to be created with a model that has not been defined
yet, it can be referenced by its name rather than the model object itself::

    from django.db import models

    class Car(models.Model):
        manufacturer = models.ForeignKey(
            "Manufacturer",
            on_delete=models.CASCADE,
        )
        # ...

> 

Author: Having introduced a `name` field in the earlier example, I think we should include it here as well, or else replace the `name` with `# ...`, or something like that, as done in other examples

Author: I'm in favour of removing the `name` field for consistency not only with the `Car` model, but also with `Manufacturer` model, which should also have a `name`, but has `# ...` to mean "other fields here"

Reviewer: Sounds good, would you like to fetch the latest and push a new revision with all changes squashed?  Thank you!

Reviewer: @cliff688 I have reviewed the latest revno available and I still see the `name = ...` definition. I understood your preference was to remove this field, did I misunderstand you? :thinking:

Author: Oh, I think when I was reviewing the changes here in patch form, the `Car` model on `FK` seemed close to the one in Lazy Relationships, so I thought we should keep them consistent, thinking they were in the same context. So, yes, you understood me (thank you), but I had misunderstood the placing of the two `Car` models, so that the issue I had imagined was there disappeared when viewing the file locally as a whole

Author: As for `Manufacturer.name` I think that's harmless.

Reviewer: Thank you @cliff688, I pushed final tweaks, I will merge when CI is green again. Amazing work! :trophy:
